### ISSUE TITLE
Unable to resolve  `.text()` on a original response after `.json()` is called on a clone

### ISSUE THREAD

**Issue by m14t** [bitinn/node-fetch#105]
Hi.  Thanks for this great library.  I really appreciate the recent addition of the `.clone` method, but believe I have found an issue with it.

Below is a test case for caching the original response, returning a cloned version, operating on that clone, and then reverting back to the original response.  While a similar test exists, this one works with Chrome's `fetch`, but times out when using `node-fetch`.

I realize that you are looking for contributors to this project and are busy with your startup, so if you have any clues on where to start, I'd be more than happy to lend a hand.

Thanks in advance!

```
	it('should allow cloning a json response and log it as text response', function(done) {
		url = base + '/json';
		var originalResponse;

		fetch(url).then(function(res) {
			// Save a copy of our untouched response to return later
			originalResponse = res;

			return res.clone();
		})
		.then(function(clonedResponse) {
			// Perform an operation that reads our cloned response's body
			var json = clonedResponse.json()

			expect(clonedResponse.bodyUsed).to.equal(true);

			return json;
		})
		.then(function(json) {
			// Use this body data
			expect(json).to.deep.equal({name: 'value'});

			return json;
		})
		.then(function(json) {
			// reset our promise chain's value to the original response, so that appears
			// as if we were never here

			return originalResponse;
		})
		.then(function(res) {
			expect(res.bodyUsed).to.equal(false);

			// Attempt to read the body a second time
			return res.text();  // BUG: THIS WILL NEVER RESOLVE
		})
		.then(function(text) {
			expect(text).to.equal('{"name":"value"}');

			done();
		});
	});
```

(cc: @kirill-konshin @bitinn)

**Comment by m14t** [comment]
I tried taking a look at this, and from what I could tell the [`self.body.on('data', ...)`](https://github.com/bitinn/node-fetch/blob/master/lib/body.js#L105) and [`self.body.on('end', ...)`](https://github.com/bitinn/node-fetch/blob/master/lib/body.js#L120) events don't seem to be getting fired when we try to call this `.text()` on the original response, which means that the returned promise from `_decode` never calls `resolve`.

Why this is happening?  I haven't got that far yet.

**Comment by bitinn** [comment]
@m14t because of the way stream works, you shouldn't use the cloned response first, use your original response first and then the cloned one.

Let me attempt to illustrate this, imagine A/B/C/D are streams:

```javascript
// when you pipe something, imagine its `data` event will be triggered sometime in future.
A.pipe(B)

// in your case, you did these:
A.pipe(B) // clone
B.pipe(C) // clonedRes.json 
A.pipe(D) // res.text

// but you did so in a way where B.pipe(C) is already done when you finally starts A.pipe(D)
// since `end` event has already fired for A, this stream has drained, thus A.pipe(D) is not working
```

There are ways we can made this easier by not allowing you to drain the original body, but instead always drain a new piped body, but I haven't sit down to work this out yet (also need to figure out issues with leaks of course)

**Comment by m14t** [comment]
Thanks @bitinn for the explanation. I've been able to use this to work around my issue.

Given that this works in the browser, should we leave this open as at least a known issue for others to find?

**Comment by bitinn** [comment]
As far as I know browser doesn't support stream yet, which is probably why they work as expected, would be interesting to see if they actually workaround this gotcha when stream spec is more or less done. But I will leave this open as it's a valid gotcha.

**Comment by bitinn** [comment]
this is now fixed and released as v1.5.3
